### ISSUE TITLE
Add a way to force usage of specific index (hints)

### ISSUE THREAD

**Issue by soualid** [arangodb/arangodb#7782]
We sometimes encounter issues with the indexes chosen by the optimizer when executing complex queries over collections containing multiple indexes. 

It would be great to extend the `cursor` and `explain` API to add an option to force the optimizer to use specific indexes (referenced by name or ID).

Actually, the only workaround we found to force the usage of specific indexes is to duplicate these collections with different indexes or to duplicate fields within these collections.

**Comment by Simran-B** [comment]
Similar request: https://github.com/arangodb/arangodb/issues/2761#issuecomment-336268514

**Comment by soualid** [comment]
Just adding a full test case for a problem we encountered recently and because it could also be a bug in the optimizer. Executing the following under versions 3.3.20 and 3.4.0 will not use the best index:

```
db._useDatabase('_system')
db._dropDatabase('test')
db._createDatabase('test')
db._useDatabase('test')
db._create('test')
db.test.ensureIndex( {"type":"skiplist","fields":["timestamp"],"unique":false,"sparse":false });
db.test.ensureIndex( {"type":"skiplist","fields":["number","timestamp"],"unique":false,"sparse":false });
db._query(`
	for i in 1..100000
	insert {
	  number: i%5,
	  timestamp: date_add(date_now(), -i%10, "minutes")
	} in test
`)
db._explain("for t in test filter t.number > 200 && t.number < 300 sort t.timestamp desc return t")
db.test.dropIndex(db.test.getIndexes()[1].id)
db._explain("for t in test filter t.number > 200 && t.number < 300 sort t.timestamp desc return t")
```

Plan chosen by the optimizer use the wrong index on `timestamp` only where an index filtering more datas exists on `["number", "timestamp"]`:

```
Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
  8   IndexNode         100000     - FOR t IN test   /* reverse skiplist index scan */
  3   CalculationNode   100000       - LET #1 = ((t.`number` > 200) && (t.`number` < 300))   /* simple expression */   /* collections used: t : test */
  4   FilterNode        100000       - FILTER #1
  7   ReturnNode        100000       - RETURN t

Indexes used:
 By   Type       Collection   Unique   Sparse   Selectivity   Fields            Ranges
  8   skiplist   test         false    false            n/a   [ `timestamp` ]   *

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-indexes
  6   use-index-for-sort
  7   remove-unnecessary-calculations-2
```

If the first index is dropped, the optimizer will chose the right index:

```
Execution plan:
 Id   NodeType           Est.   Comment
  1   SingletonNode         1   * ROOT
  8   IndexNode         13333     - FOR t IN test   /* skiplist index scan */
  5   CalculationNode   13333       - LET #3 = t.`timestamp`   /* attribute expression */   /* collections used: t : test */
  6   SortNode          13333       - SORT #3 DESC
  7   ReturnNode        13333       - RETURN t

Indexes used:
 By   Type       Collection   Unique   Sparse   Selectivity   Fields                      Ranges
  8   skiplist   test         false    false            n/a   [ `number`, `timestamp` ]   ((t.`number` > 200) && (t.`number` < 300))

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-indexes
  6   remove-filter-covered-by-index
  7   remove-unnecessary-calculations-2
```

Considering this, we had no other choice than duplicating the `timestamp` value to force the optimizer to use the proper index, a hint passed to the optimizer through option or directly in the AQL query could be really helpfull in such cases.

Thanks!

**Comment by jsteemann** [comment]
@soualid : I agree there should be a way to tell the optimizer to prefer a specific index.
However, for these particular queries, I think the optimizer is not necessarily wrong.
The query is:
```
FOR t IN test 
  FILTER t.number > 200 && t.number < 300 
  SORT t.timestamp desc 
  RETURN t
```
So it filters on a range and also sorts by a different attribute. This requires the index to be sorted by `number` and `timestamp` at the same time, which is not how our indexes work.

Here is an example for an index created on attributes `number` and `timestamp` (in this very order):
```
number       timestamp
   200               1
   200               2
   200               3
   201               1
   201               2
   201               3
   202               1
   202               2
   202               3
   202               4
   210               1
   ... 
```
As you can see, the index is primarily sorted by `number` here, so a range scan is easily possible. However, range between number `200` and `300` will be sorted by `number` first, and only then `timestamp`. The query however requires the result set to be sorted by `timestamp` (not `number`, `timestamp`).
Turning the index attributes around in the index will not help either, because that index would be sorted by `timestamp`, but for each distinct `timestamp` value in the index there would be disjunct ranges the desired `number` range that would need to be extracted.

Given the two indexes present in your case, the choice for the optimizer is to either use the index on `timestamp` for retrieving values in `timestamp` order and then post-filtering all non-matching `number` values **or** to use the index on `number` and `timestamp` for filtering on all matching numbers using the index, and then doing an extra post-sort to sort the result by `timestamp` values descending.
But none of the two indexes will allow filtering and sorting at the same time.

For this particular query the optimizer chose that the sorting may be more expensive than filtering, so it has used the index which supports sorting by `timestamp`. That decision is based on heuristics, and it may be non-optimal. When dropping that index, the optimizer will only have one index available, which it will use for the filter condition. That is the index on `number` and `timestamp`. However, the query will only use the `number` part of the index here, and ignore the `timestamp` part. As can be seen in the execution plan, the query still needs to do a post-sort when using that combined index. So even if the index is created on two attributes, only the first index attribute is actually used in the query.

Long story short: I don't think there is a way to use a combined index on `number` and `timestamp` (in whatever order) here.

That being said, I think a sensible way of preferring an index still will make a lot of sense here.

**Comment by soualid** [comment]
Very clear, many thanks for taking the time to explain internals here. My test was skewed because I thought that sorting on the secondary field of an index was possible or considered more performant by the optimizer. 

Having a better understanding of this, I agree that on a single field, the optimizer has no way to know that more datas would be filtered using the `number` index over the `timestamp` one without analyzing the underlying datas, which is definitely not feasible during the optimization phase.

Considering most applications need different indexes for different use cases on the same datas, giving users a way to force the engine to use a specific index for a particular use case sounds like an important feature. I hope it will be considered for implementation in a near future since duplicating indexed fields is a working workaround, but it feels a bit hacky and it is not optimized for large collections.

Thank you again for the explanation, feel free to close this ticket if #2761 already stands for the feature request.

**Comment by jsteemann** [comment]
Hi @soualid : I personally prefer to leave this ticket open, to increase the pressure for finally adding a "force index"/"prefer index" feature.

**Comment by Yogu** [comment]
Is there some documentation or a tool that shows how ArangoDB decides whether to use an index for sorting or one for filtering, if both would be possible? `explain` does not really explain this decision - it only shows the index it finally chose. Using `allPlans` also does not help because I've never seen more than one plan returned by it.

**Comment by jsteemann** [comment]
@yogu: `allPlans` is an option for the HTTP REST explain API `/_api/explain`. This API supports returning multiple plans, however, `db._explain()` in the arangosh or in the web UI will always print just a single execution plan, namely the one with the lowest estimated cost.
I think it makes sense to have db._explain() optionally show all plans.

However, that alone will not help much to shed more light on the index selection decisions, because that info is not exposed anywhere in the internal explain results and all the influencing factors are not carried through and preserved after the decision was taken. That would require some completely new structures to pass around internally for all the decisions the optimizer has made, so I doubt it is as simply as it may sound.
There are many factors that influence the optimizer decision for indexes, e.g. index uniqueness and sparsity, how many documents that are in the collection, number of attributes covered by the index, the type of comparison (equality lookup, range scan), type of index, index selectivity estimate (if present) etc. The rules are currently heuristics internalized in the code, which are adjusted from time to time (i.e. there can be changes from release to release).
Making these decisions transparent sounds good, but as I said it will not be trivial and thus may not be done in the very foreseeable future. What makes more sense from point of view now (because it sounds more achievable) is to provide some syntax for index usage hints.

**Comment by Yogu** [comment]
@jsteemann Yes I agree that the `allPlans` is pretty cumbersome to use. It would be great to include it in the UI and in `explainer.js`. I solved it via JavaScript in arangosh and found that `plans.length` is `1`. By the way, it would also be good if the `query` method would accept `allPlans` so that the *profile* feature (explaining + executing) could make use of it.

I see how explaining the index decision is not trivial, and I guess if that will be added as a feature then it's going to be on the long term. Considering it's all based around herustics that might be hard to adjust from case to case anyway, the feature suggested in this issue here - enabling to force a specific index - would maybe be of more use in the mid-term.

**Comment by jsteemann** [comment]
@yogu: adjusting the explainer to show multiple plans is probably trivial... I have quickly prototyped the change:
```
diff --git a/js/common/modules/@arangodb/aql/explainer.js b/js/common/modules/@arangodb/aql/explainer.js
index fdf1c749a0..f37392489e 100644
--- a/js/common/modules/@arangodb/aql/explainer.js
+++ b/js/common/modules/@arangodb/aql/explainer.js
@@ -40,15 +40,15 @@ const anonymize = function(doc) {
   return doc;
 };
 
-var stringBuilder = {
+let stringBuilder = {
   output: '',
+  prefix: '',
 
   appendLine: function (line) {
-    if (!line) {
-      this.output += '\n';
-    } else {
-      this.output += line + '\n';
+    if (line) {
+      this.output += this.prefix + line;
     }
+    this.output += '\n';
   },
 
   getOutput: function () {
@@ -57,6 +57,11 @@ var stringBuilder = {
 
   clearOutput: function () {
     this.output = '';
+  },
+
+  wrap: function (str, width) {
+    let re = '.{1,' + width + '}(\\s|$)|\\S+?(\\s|$)';
+    return str.match(new RegExp(re, 'g')).join('\n' + this.prefix).replace(/\n+/g, '\n ' + this.prefix);
   }
 
 };
@@ -154,12 +159,6 @@ function pad (n) {
   return new Array(n).join(' ');
 }
 
-function wrap (str, width) {
-  'use strict';
-  var re = '.{1,' + width + '}(\\s|$)|\\S+?(\\s|$)';
-  return str.match(new RegExp(re, 'g')).join('\n');
-}
-
 /* print functions */
 
 /* print query string */
@@ -174,7 +173,7 @@ function printQuery (query) {
   } else {
     stringBuilder.appendLine(section('Query String:'));
   }
-  stringBuilder.appendLine(' ' + value(wrap(query, 100).replace(/\n+/g, '\n ', query)));
+  stringBuilder.appendLine(' ' + value(stringBuilder.wrap(query, 100)));
   stringBuilder.appendLine();
 }
 
@@ -605,7 +604,7 @@ function printShortestPathDetails (shortestPaths) {
 }
 
 /* analyze and print execution plan */
-function processQuery (query, explain) {
+function processQuery (query, explain, planIndex) {
   'use strict';
   var nodes = { },
     parents = { },
@@ -617,8 +616,12 @@ function processQuery (query, explain) {
     maxCallsLen = String('Calls').length,
     maxItemsLen = String('Items').length,
     maxRuntimeLen = String('Runtime [s]').length,
-    plan = explain.plan,
     stats = explain.stats;
+    
+  let plan = explain.plan;
+  if (planIndex !== undefined) {
+    plan = explain.plans[planIndex];
+  }
 
   /// mode with actual runtime stats per node
   let profileMode = stats && stats.hasOwnProperty('nodes');
@@ -1598,8 +1601,23 @@ function explain(data, options, shouldPrint) {
 
   stringBuilder.clearOutput();
   let stmt = db._createStatement(data);
-  let result = stmt.explain(options); // TODO why is this there ?
-  processQuery(data.query, result);
+  let result = stmt.explain(options); 
+  if (options.allPlans) {
+    // multiple plans
+    for (let i = 0; i < result.plans.length; ++i) {
+      if (i > 0) {
+        stringBuilder.appendLine();
+      }
+      stringBuilder.appendLine(section("Plan #" + (i + 1) + " (estimated cost: " + result.plans[i].estimatedCost + ")"));
+      stringBuilder.prefix = ' ';
+      stringBuilder.appendLine();
+      processQuery(data.query, result, i);
+      stringBuilder.prefix = '';
+    }
+  } else {
+    // single plan
+    processQuery(data.query, result, undefined);
+  }
 
   if (shouldPrint === undefined || shouldPrint) {
     print(stringBuilder.getOutput());
@@ -1617,13 +1635,14 @@ function profileQuery(data, shouldPrint) {
   }
   let options =  data.options || { };
   options.silent = true;
+  options.allPlans = false; // always turn this off, as it will not work with profiling
   setColors(options.colors === undefined ? true : options.colors);
 
   stringBuilder.clearOutput();
   let stmt = db._createStatement(data);
   let cursor = stmt.execute();
   let extra = cursor.getExtra();
-  processQuery(data.query, extra);
+  processQuery(data.query, extra, undefined);
 
   if (shouldPrint === undefined || shouldPrint) {
     print(stringBuilder.getOutput());
```
This will work from the shell already, but will not add the feature to the web interface yet. I have yet to figure out how/if this can be done.
And hinting for using a specific index is also a feature we want to provide in one of the upcoming releases. But it still has to be specified and then implemented, so please don't expect it too soon.

**Comment by soualid** [comment]
Just wanted to put a new coin on this feature request as we just had a little surprise in production that this feature could prevent.

On a large collection of our cluster (~22M documents), the underlying datas ventilation changed which made the optimizer to brutally decide to use another index for some queries using this collection, which resulted in a dramatical performance degradation on these – unmodified – queries.

Having a way to force which index to use per query could improve stability and prevent such problems in production.

Considering the workarounds are almost always "hacky" and may require reindexing or manipulating the underlying datas, which cannot be achieved without downtime on big collections. This really looks like a feature to prioritize.

Thanks, and sorry for upping this issue!

**Comment by ifcologne** [comment]
Hi Simon, 
we did some investigations earlier this year and have picked a proposal for implementing in one of the next releases. 

We need to introduce named indexes first and add support for an option IndexHints. There is various of complexity one could add to specify hints and we will start with the simplest one, which means you can specify a single index hint. 

`OPTIONS {indexHint: 'byName'}`

Later you might specify more than one hint or collection specific index hints. Work has not yet started, but it has some priority in our backlog.

**Comment by maxkernbach** [comment]
Hi @soualid,

both named indexes and the option "indexHint" are supported in Release Candidate 7 of ArangoDB 3.5.

For further details, please have a look at:

- [Features and Improvements in ArangoDB 3.5 - Named indices](https://www.arangodb.com/docs/devel/release-notes-new-features35.html#named-indices)
- [Working with Indexes documentation](https://www.arangodb.com/docs/devel/indexing-working-with-indexes.html)
- [Features and Improvements in ArangoDB 3.5 - Index Hints](https://www.arangodb.com/docs/devel/release-notes-new-features35.html#index-hints-in-aql)
- [Index Hints documentation](https://www.arangodb.com/docs/devel/aql/operations-for.html#index-hints)
- [Index Hints tutorial](https://www.arangodb.com/arangodb-training-center/index-hints-named-indices/)


You can find the latest Release Candidate of 3.5 in the [download section](https://www.arangodb.com/download-technical-preview/).
